src/modules/frost/session_impl.h

```diff
@@ -0,0 +1,646 @@
+    unsigned char *sk_ser_ptr = NULL;
+    int ret = 1;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secnonce != NULL);
+    memset(secnonce, 0, sizeof(*secnonce));
+    ARG_CHECK(pubnonce != NULL);
+    memset(pubnonce, 0, sizeof(*pubnonce));
+    ARG_CHECK(session_id32 != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    if (agg_share == NULL) {
+        /* Check in constant time that the session_id is not 0 as a
+         * defense-in-depth measure that may protect against a faulty RNG. */
+        unsigned char acc = 0;
+        for (i = 0; i < 32; i++) {
+            acc |= session_id32[i];
+        }
+        ret &= !!acc;
+        memset(&acc, 0, sizeof(acc));
+    }
+
+    if (agg_share != NULL) {
+        /* Check that the agg_share is valid to be able to sign for it later. */
+        secp256k1_scalar sk;
+        int ret_tmp;
+
+        ret &= secp256k1_frost_share_load(ctx, &sk, agg_share);
+        secp256k1_scalar_clear(&sk);
+
+        ret_tmp = secp256k1_frost_share_serialize(ctx, sk_ser, agg_share);
+        VERIFY_CHECK(ret_tmp);
+        sk_ser_ptr = sk_ser;
+    }
+
+    if (agg_pk != NULL) {
+        if (!secp256k1_xonly_pubkey_serialize(ctx, pk_ser, agg_pk)) {
+            return 0;
+        }
+        pk_ser_ptr = pk_ser;
+    }
+    secp256k1_nonce_function_frost(k, session_id32, msg32, sk_ser_ptr, pk_ser_ptr, extra_input32);
+    VERIFY_CHECK(!secp256k1_scalar_is_zero(&k[0]));
+    VERIFY_CHECK(!secp256k1_scalar_is_zero(&k[1]));
+    VERIFY_CHECK(!secp256k1_scalar_eq(&k[0], &k[1]));
+    secp256k1_frost_secnonce_save(secnonce, k);
+    secp256k1_frost_secnonce_invalidate(ctx, secnonce, !ret);
+
+    for (i = 0; i < 2; i++) {
+        secp256k1_gej nonce_ptj;
+        secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &nonce_ptj, &k[i]);
+        secp256k1_ge_set_gej(&nonce_pt[i], &nonce_ptj);
+        secp256k1_declassify(ctx, &nonce_pt[i], sizeof(nonce_pt));
+        secp256k1_scalar_clear(&k[i]);
+    }
+    /* nonce_pt won't be infinity because k != 0 with overwhelming probability */
+    secp256k1_frost_pubnonce_save(pubnonce, nonce_pt);
+    return ret;
+}
+
+static int secp256k1_frost_sum_nonces(const secp256k1_context* ctx, secp256k1_gej *summed_nonces, const secp256k1_frost_pubnonce * const* pubnonces, size_t n_pubnonces) {
+    size_t i;
+    int j;
+
+    secp256k1_gej_set_infinity(&summed_nonces[0]);
+    secp256k1_gej_set_infinity(&summed_nonces[1]);
+
+    for (i = 0; i < n_pubnonces; i++) {
+        secp256k1_ge nonce_pt[2];
+        if (!secp256k1_frost_pubnonce_load(ctx, nonce_pt, pubnonces[i])) {
+            return 0;
+        }
+        for (j = 0; j < 2; j++) {
+            secp256k1_gej_add_ge_var(&summed_nonces[j], &summed_nonces[j], &nonce_pt[j], NULL);
+        }
+    }
+    return 1;
+}
+
+/* Implements binding factor from draft-irtf-cfrg-frost-04, section 4.4. */
+/* TODO: consider updating to frost-08 to address maleability at the cost of performance */
```

Comment:
If this is not done and a single nonce coefficient `b` is used for all signers, then the input to the hashing function that determines `b` must include all values that affect the value of Lagrange coefficient `l`. Otherwise, the implementation is vulnerable to an attack on `l` that is similar to the attack against two round musig without nonce coefficients.

Similar to the implication `c != c' => b != b'` (where `b,c` and `b',c'` are the values for the nonce coefficients and challenges in two execution of the signing algorithm) mentioned here, we want `l != l' => b != b'`. In particular this means that the set of signers participating in the signature needs to be hashed into `b`.

Reply:
>then the input to the hashing function that determines b must include all values that affect the value of Lagrange coefficient l

The participant indexes (the `indexhash`s) are the only values that affect the Lagrange coefficient, so I believe if we add them to the `noncehash` that will ensure when `l` changes, so does `b`.

Reply:
Looking back at https://datatracker.ietf.org/doc/draft-irtf-cfrg-frost/04/, the omission of the participant index is an implmentation error on my part. The rfc specifies:

```
def encode_group_commitment_list(commitment_list):
    encoded_group_commitment = nil
    for (index, hiding_nonce_commitment, binding_nonce_commitment) in commitment_list:
      encoded_commitment = encode_uint16(index) ||
                           G.SerializeElement(hiding_nonce_commitment) ||
                           G.SerializeElement(binding_nonce_commitment)
      encoded_group_commitment = encoded_group_commitment || encoded_commitment
    return encoded_group_commitment

def compute_binding_factor(encoded_commitment_list, msg):
       msg_hash = H3(msg)
       rho_input = encoded_commitment_list || msg_hash
       binding_factor = H1(rho_input)
       return binding_factor
```